activemodel/lib/active_model/error.rb

```diff
@@ -0,0 +1,197 @@
+# frozen_string_literal: true
+
+module ActiveModel
+  # == Active \Model \Error
+  #
+  # Represents one single error
+  class Error
+    def initialize(base, attribute, type = nil, **options)
+      @base = base
+      @attribute = attribute
+      @type = type || :invalid
+      @options = options
+    end
+
+    def initialize_dup(other)
+      @attribute = @attribute.dup
+      @type = @type.dup
+      @options = @options.deep_dup
+    end
+
+    attr_reader :base, :attribute, :type, :options
+
+    # Translates an error message in its default scope
+    # (<tt>activemodel.errors.messages</tt>).
+    #
+    # Error messages are first looked up in <tt>activemodel.errors.models.MODEL.attributes.ATTRIBUTE.MESSAGE</tt>,
+    # if it's not there, it's looked up in <tt>activemodel.errors.models.MODEL.MESSAGE</tt> and if
+    # that is not there also, it returns the translation of the default message
+    # (e.g. <tt>activemodel.errors.messages.MESSAGE</tt>). The translated model
+    # name, translated attribute name and the value are available for
+    # interpolation.
+    #
+    # When using inheritance in your models, it will check all the inherited
+    # models too, but only if the model itself hasn't been found. Say you have
+    # <tt>class Admin < User; end</tt> and you wanted the translation for
+    # the <tt>:blank</tt> error message for the <tt>title</tt> attribute,
+    # it looks for these translations:
+    #
+    # * <tt>activemodel.errors.models.admin.attributes.title.blank</tt>
+    # * <tt>activemodel.errors.models.admin.blank</tt>
+    # * <tt>activemodel.errors.models.user.attributes.title.blank</tt>
+    # * <tt>activemodel.errors.models.user.blank</tt>
+    # * any default you provided through the +options+ hash (in the <tt>activemodel.errors</tt> scope)
+    # * <tt>activemodel.errors.messages.blank</tt>
+    # * <tt>errors.attributes.title.blank</tt>
+    # * <tt>errors.messages.blank</tt>
+    def message
+      type = @type
+      if msg = @options[:message]
+        if msg.is_a?(Symbol)
+          # TODO: Having options[:message] to act as an alternative i18n lookup key seems to be wrong.
+          # We should have a separate new key for this, for example `:lookup_key`,
+          # this way lookup_key can also be a Proc and generated dynamically.
+          type = msg
+          msg = nil
+        end
```

Comment:
Should not this method return `msg` if `options[:message]` is a string? I see no reason to do i18n lookup scope if we already passed a string.

Reply:
This was not happening before. Because this change I'm getting `undefined method `model_name' for nil:NilClass` in my application for errors created with `Errors.new(nil)` but with errors being added as full strings without using i18n. Before I think it was not passing through the i18n layer because this exception is not raised in the current implementation.

Reply:
My bad, I think you are right. I don't recall the reason but I omitted the logic in the original `normalize_message`. I've added the return if `type` is not a symbol.